osgar/explore.py

```diff
@@ -0,0 +1,146 @@
+"""
+  Explore - follow a wall/obstacles
+    (use 2D SICK LIDAR only)
+"""
+import math
+from datetime import timedelta
+
+import numpy as np
+
+from osgar.node import Node
+
+
+SAFE_DISTANCE_STOP = 0.5
+SAFE_DISTANCE_GO = SAFE_DISTANCE_STOP + 0.3
+WALL_DISTANCE = 0.7  # m
+DESIRED_SPEED = 0.4  # m/s
+
+
+def min_dist(data):
+    data = np.array(data)
+    mask = (data > 100)
+    if np.any(mask):
+        return np.min(data[mask]) * 0.001
+    return None 
+
+
+def min_dist_arr(data):
+    """Fake segments as in demo.py for Velodyne"""
+    num = len(data)
+    # laser data are anticlockwise -> swap left, right
+    return min_dist(data[num//2:]), min_dist(data[:num//2])
+
+
+def tangent_circle(dist, radius):
+    if dist < 2 * radius:
+        if dist >= radius:
+            return math.asin(radius/float(dist))
+        return math.radians(100)
+    return None
+
+
+def follow_wall_angle(laser_data, radius, right_wall=False):
```

Comment:
I need some description of the inputs and outputs.

Reply:
There is not much I can do to review this function since I don't understand the inputs and also I am not sure what exactly the output should be.

Reply:
Ok, so it finds the angle to the closest point in laser scan (either on the left or right side). Then calculates an angle to a free space. This angle is returned. But somehow it is used as angular speed. Is this observation correct?

Reply:
correct - used for function description